docs/predictive-analytics/next-best-offer/index.md

```diff
@@ -0,0 +1,208 @@
+
+Create the table for customer information:
+
+```sql
+CREATE TABLE customers (
+    CUSTOMER_ID INTEGER PRIMARY KEY,
+    FIRST_NAME STRING,
+    LAST_NAME STRING,
+    EMAIL STRING,
+    GENDER STRING,
+    INCOME INTEGER,
+    FICO INTEGER
+) WITH (
+    KAFKA_TOPIC = 'CUSTOMERS_TABLE',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+```
+
+Typically, customer information would be sourced from an existing database. As customer details change, tables in the database are updated and we can stream them into Kafka using Kafka Connect with [change data capture](https://www.confluent.io/blog/cdc-and-streaming-analytics-using-debezium-kafka/) (CDC).  The primary key for the `customers` table is the customer id, which corresponds to the key of the `customer_activity_stream`. This stream facilitates joins for enriching customer information. In this example, we'll populate the `customers` table by executing the following `INSERT` statements:
+
+```sql
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (1,'Waylen','Tubble','wtubble0@hc360.com','Male',403646, 465);
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (2,'Joell','Wilshin','jwilshin1@yellowpages.com','Female',109825, 624);
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (3,'Ilaire','Latus','ilatus2@baidu.com','Male',407964, 683);
+```
+
+#### Creating the offer table
+
+The last lookup table you'll add is the `OFFERS` table:
+
+```sql
+CREATE TABLE offers (
+    OFFER_ID INTEGER PRIMARY KEY,
+    OFFER_NAME STRING,
+    OFFER_URL STRING
+) WITH (
+    KAFKA_TOPIC = 'OFFERS_STREAM',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+```
+
+This table provides the enrichment information that will be needed when the application calculates the next best offer.
+
+To populate the `OFFERS` table, execute the following `INSERT` statements:
+
+```sql
+INSERT INTO offers (offer_id, offer_name, offer_url) VALUES (1,'new_savings','http://google.com.br/magnis/dis/parturient.json');
+INSERT INTO offers (offer_id, offer_name, offer_url) VALUES (2,'new_checking','https://earthlink.net/in/ante.js');
+INSERT INTO offers (offer_id, offer_name, offer_url) VALUES (3,'new_home_loan','https://webs.com/in/ante.jpg');
+INSERT INTO offers (offer_id, offer_name, offer_url) VALUES (4,'new_auto_loan','http://squidoo.com/venenatis/non/sodales/sed/tincidunt/eu.js');
+INSERT INTO offers (offer_id, offer_name, offer_url) VALUES (5,'no_offer','https://ezinearticles.com/ipsum/primis/in/faucibus/orci/luctus.html');
+```
+
+### Determining the next best offer
+
+Now you'll create the event stream that calculates the next best offer for your customers based on their activity. 
+
+#### Calculating the offer
+
+To perform the next offer calculation, create a stream that performs a join between the `CUSTOMER_ACTIVITY_STREAM` and the `CUSTOMERS` table:
+
+```sql
```

Comment:
Above we create the input stream first and insert value into it, before we create the table and add data. This implies that the stream data will have lower timestamps compared to the table data and thus would join agains an "empty" table, as it will be processed first. -- Should we switch both above to ensure that the stream-table joins works as expected?

It also applies to the "offer" table, ie, the input STREAM should be created (well, populated with data would be sufficient) last?

Reply:
Yeah I overlooked that aspect when building this - I've updated to create the tables first as well as insert the data